src/puppetlabs/puppetdb/cli/services.clj

```diff
@@ -106,18 +134,22 @@
     (catch Exception e
       (log/error e (trs "Error while deactivating stale nodes")))))
 
-(defn purge-nodes!
-  "Delete nodes which have been *deactivated or expired* longer than
-  `node-purge-ttl`."
-  [node-purge-ttl db]
-  {:pre [(map? db)
-         (period? node-purge-ttl)]}
+(defn-validated purge-nodes!
+  "Deletes nodes which have been *deactivated or expired* longer than
+  node-purge-ttl.  Deletes at most batch_limit nodes if opts is a map,
+  all relevant nodes otherwise."
+  [node-purge-ttl :- (s/pred period?)
+   opts :- (s/if map? purge-nodes-opts-schema (s/eq true))
+   db :- (s/pred map?)]
   (try
     (kitchensink/demarcate
      (format "purge deactivated and expired nodes (threshold: %s)"
              (format-period node-purge-ttl))
-     (jdbc/with-transacted-connection db
-       (scf-store/purge-deactivated-and-expired-nodes! (ago node-purge-ttl))))
+     (let [horizon (ago node-purge-ttl)]
+       (jdbc/with-transacted-connection db
+         (if-let [limit (and (map? opts) (:batch_limit opts))]
```

Comment:
See the (arg) schema - opts can be `true`, or a map of options, where `true` means "do the default thing" (i.e. no batch limit) -- that's due to the way we handle the various commands in the caller (`clean-up`), but could of course be changed.